core/lib/src/fs/server.rs

```diff
@@ -60,351 +72,492 @@ use crate::fs::NamedFile;
-    /// directory, if it exists.
-    ///
-    /// When enabled, [`FileServer`] will respond to requests for a directory
-    /// `/foo` or `/foo/` with the file at `${root}/foo/index.html` if it
-    /// exists. When disabled, requests to directories will always forward.
-    ///
-    /// **Enabled by default.**
-    pub const Index: Options = Options(1 << 0);
+/// Prepends the provided path, to serve a single static file.
+///
+/// # Example
+///
+/// ```rust,no_run
+/// # use rocket::fs::{FileServer, file_root};
+/// # fn make_server() -> FileServer {
+/// FileServer::empty()
+///     .map_file(file_root("static/index.html"))
+/// # }
+/// ```
+///
+/// # Panics
+///
+/// Panics if `path` does not exist.
+pub fn file_root(path: impl AsRef<Path>)
+    -> impl for<'p, 'h> Fn(File<'p, 'h>) -> FileResponse<'p, 'h> + Send + Sync + 'static
+{
+    use yansi::Paint as _;
 
-    /// Allow serving dotfiles.
-    ///
-    /// When enabled, [`FileServer`] will respond to requests for files or
-    /// directories beginning with `.`. When disabled, any dotfiles will be
-    /// treated as missing.
-    ///
-    /// **Disabled by default.**
-    pub const DotFiles: Options = Options(1 << 1);
+    let path = path.as_ref();
+    if !path.exists() {
+        let path = path.display();
+        error!("FileServer path '{}' is not a file.", path.primary());
+        warn!("Aborting early to prevent inevitable handler error.");
+        panic!("invalid file: refusing to continue");
+    }
+    let path = path.to_path_buf();
+    move |f| {
+        FileResponse::File(f.map_path(|p| path.join(p)))
+    }
+}
 
-    /// Normalizes directory requests by redirecting requests to directory paths
-    /// without a trailing slash to ones with a trailing slash.
-    ///
-    /// **Enabled by default.**
+/// Prepends the provided path, without checking to ensure the path exists during
+/// startup.
+///
+/// # Example
+///
+/// ```rust,no_run
+/// # use rocket::fs::{FileServer, missing_root};
+/// # fn make_server() -> FileServer {
+/// FileServer::empty()
+///     .map_file(missing_root("/tmp/rocket"))
+/// # }
+/// ```
+pub fn missing_root(path: impl AsRef<Path>)
+    -> impl for<'p, 'h> Fn(File<'p, 'h>) -> FileResponse<'p, 'h> + Send + Sync + 'static
+{
+    let path = path.as_ref().to_path_buf();
+    move |f| {
+        FileResponse::File(f.map_path(|p| path.join(p)))
+    }
+}
+
+/// Filters out any path that contains a file or directory name starting with a
+/// dot. If used after `dir_root`, this will also check the root path for dots, and
+/// filter them.
+///
+/// # Example
+///
+/// ```rust,no_run
+/// # use rocket::fs::{FileServer, filter_dotfiles, dir_root};
+/// # fn make_server() -> FileServer {
+/// FileServer::empty()
+///     .filter_file(filter_dotfiles)
+///     .map_file(dir_root("static"))
+/// # }
+/// ```
+pub fn filter_dotfiles(file: &File<'_, '_>) -> bool {
+    !file.path.iter().any(|s| s.as_encoded_bytes().starts_with(b"."))
```

Comment:
This will filter out things like `foo/../bar` and `foo/./bar` which are totally okay and safe. We need to allow this paths. We should have tests to ensure that we do allow those paths. I fixed the latter case in my [patch](https://paste.rs/AbR41.patch).

Reply:
This actually won't filter out `..`, since they are already stripped out by `FileServer` when it calls `to_path_buf` on the URI. I think we should modify `to_path_buf` to filter out `.` segments in a similar way (by just ignoring them).